src/python/pants/core/goals/test.py

```diff
@@ -533,53 +744,69 @@ async def run_tests(
     union_membership: UnionMembership,
     distdir: DistDir,
     run_id: RunId,
+    specs: Specs,
 ) -> Test:
-    if test_subsystem.debug or test_subsystem.debug_adapter:
-        return await _run_debug_tests(test_subsystem, debug_adapter)
-
-    shard, num_shards = parse_shard_spec(test_subsystem.shard, "the [test].shard option")
-    targets_to_valid_field_sets = await Get(
-        TargetRootsToFieldSets,
-        TargetRootsToFieldSetsRequest(
-            TestFieldSet,
-            goal_description=f"the `{test_subsystem.name}` goal",
-            no_applicable_targets_behavior=NoApplicableTargetsBehavior.warn,
-            shard=shard,
-            num_shards=num_shards,
-        ),
+    request_types = union_membership.get(TestRequest)
+    partitioners = union_membership.get(TestRequest.PartitionRequest)
+    partitions_by_request_type = await _get_partitions_by_request_type(
+        request_types, partitioners, specs, test_subsystem
     )
-    environment_names_per_field_set = await MultiGet(
-        Get(
-            EnvironmentName,
-            EnvironmentNameRequest,
-            EnvironmentNameRequest.from_field_set(field_set),
+    # TODO: Restore warning/error if no applicable targets were matched.
+
+    def batch(
+        iterable: Iterable[_T], key: Callable[[_T], str] = lambda x: str(x)
+    ) -> Iterator[tuple[_T, ...]]:
+        batches = partition_sequentially(
+            iterable,
+            key=key,
+            size_target=test_subsystem.batch_size,
+            size_max=4 * test_subsystem.batch_size,
         )
-        for field_set in targets_to_valid_field_sets.field_sets
-    )
+        for batch in batches:
+            yield tuple(batch)
+
+    batches_by_request_type = {
+        request_type: [
+            (subpartition, key)
+            for partitions in partitions_list
+            for key, partition in partitions.items()
+            for subpartition in batch(partition)
+        ]
+        for request_type, partitions_list in partitions_by_request_type.items()
+    }
+
+    subpartitions = [
+        request_type.SubPartition(elements, key)
+        for request_type, batch in batches_by_request_type.items()
+        for elements, key in batch
+    ]
+
+    if test_subsystem.debug or test_subsystem.debug_adapter:
+        return await _run_debug_tests(subpartitions, test_subsystem, debug_adapter)
+
+    # TODO: Environments??
```

Comment:
@Eric-Arellano I'm thinking `EnvironmentName` should be another field on `Partitions` and `TestRequest.SubPartition`. The downside is that each plugin will need to implement the environment lookup, but we can hopefully cover many plugins by pushing the logic into `Partitions.partition_per_input` - does that sound plausible?

Reply:
We need to use the same environment for the whole partition. Correctly partitioning by `environment` probably belongs inside each partition rule, just like they need to consider `resolve`. But we can use rule helpers to DRY stuff

I'm thinking in `test.py` we could do one of these two:

1. Choose an arbitrary element of the partitions and determine the environment based on that. Trust that the others are good to go.
2. Determine the environment for each field set in the partition, and raise an AssertionError if they are not all identical because it represents a programmer error in their partition rule

The second is certainly a more developer-friendly approach and will avoid common bugs.

Reply:
I've pushed an implementation of option 2